src/Microsoft.DotNet.Build.Tasks/FilterRuntimesFromSupports.cs

```diff
@@ -0,0 +1,89 @@
+                tfmRidPairs.Add(compatibilityProfile);
+            }
+            return tfmRidPairs;
+        }
+
+        public static List<CompatibilityProfile> FilterForApplicableTFMRIDPairs(List<CompatibilityProfile> tfmRidPairs,
+            string framework, string runtimeIdentifier)
+        {
+            List<CompatibilityProfile> filteredSupports = new List<CompatibilityProfile>();
+            List<CompatibilityProfile> filteredRuntimes = new List<CompatibilityProfile>();
+            foreach (var compatibilityProfile in tfmRidPairs)
+            {
+                foreach (var tfmRidPair in compatibilityProfile.RestoreContexts)
+                {
+                    if (NuGetFramework.Parse(framework).Equals(tfmRidPair.Framework))
+                    {
+                        filteredSupports.Add(compatibilityProfile);
+                        break;
+                    }
+                }
+            }
+
+            foreach (var compatibilityProfile in filteredSupports)
+            {
+                for (int i = compatibilityProfile.RestoreContexts.Count - 1; i >= 0; i--)
+                {
+                    if (!runtimeIdentifier.Equals(compatibilityProfile.RestoreContexts[i].RuntimeIdentifier))
+                    {
+                        compatibilityProfile.RestoreContexts.RemoveAt(i);
+                    }
+                }
+                if (compatibilityProfile.RestoreContexts.Count > 0)
+                {
+                    filteredRuntimes.Add(compatibilityProfile);
+                }
+            }
+
+            return filteredRuntimes;
+        }
+
+        public static JObject GenerateCustomSupportsClause(List<CompatibilityProfile> applicableTfmRidPairs)
+        { 
+            JObject supportsClause = new JObject();
+            foreach (var compatibilityProfile in applicableTfmRidPairs)
+            {
+                Dictionary<string, List<string>> map = new Dictionary<string, List<string>>();
+                foreach (var tfmRidPair in compatibilityProfile.RestoreContexts)
+                {
+                    if (map.ContainsKey(tfmRidPair.Framework.GetShortFolderName()))
+                    {
+                        List<string> str = new List<string>();
+                        map.TryGetValue(tfmRidPair.Framework.GetShortFolderName(), out str);
+                        str.Add(tfmRidPair.RuntimeIdentifier);
+                    }
+                    else
+                    {
+                        map.Add(tfmRidPair.Framework.GetShortFolderName(),
+                            new List<string>() {tfmRidPair.RuntimeIdentifier});
+                    }
+                }
+                var tfmRidsForSupports = new JObject();
+                foreach (var item in map)
+                {
+                    tfmRidsForSupports[item.Key] = new JArray(item.Value);
+                }
+                supportsClause[compatibilityProfile.Name] = tfmRidsForSupports;
```

Comment:
It might be cleaner if we just use the TFM:RID tuples and put them under a single inline supports clause, rather than preserving the compatibilityProfile names and subsetting them.  I don't trust NuGet to handle conflicts correctly and at a minimum it is confusing.